src/bindings/rust/elektra/src/kdb.rs

```diff
@@ -0,0 +1,405 @@
+    pub fn set<'a>(
+        &mut self,
+        keyset: &mut KeySet,
+        key: &mut StringKey<'a>,
+    ) -> Result<bool, KDBError<'a>> {
+        let ret_val = unsafe { elektra_sys::kdbSet(self.as_ptr(), keyset.as_ptr(), key.as_ptr()) };
+
+        if ret_val == 1 {
+            Ok(true)
+        } else if ret_val == 0 {
+            Ok(false)
+        } else {
+            Err(map_kdb_error(key))
+        }
+    }
+
+    /// This method can be used the given KDB handle meets certain clauses, specified in contract.
+    /// The return value is true on success,
+    /// and false if clauses of the contract are unmet.
+    pub fn ensure<'a>(
+        &mut self,
+        keyset: &mut KeySet,
+        key: &mut StringKey<'a>,
+    ) -> Result<bool, KDBError<'a>> {
+        let ret_val =
+            unsafe { elektra_sys::kdbEnsure(self.as_ptr(), keyset.as_ptr(), key.as_ptr()) };
+
+        if ret_val == 0 {
+            Ok(true)
+        } else if ret_val == 1 {
+            Ok(false)
+        } else {
+            Err(map_kdb_error(key))
+        }
+    }
+
+    pub fn as_ptr(&mut self) -> *mut elektra_sys::KDB {
+        self.ptr.as_ptr()
+    }
+}
+
+impl AsRef<elektra_sys::KDB> for KDB {
+    fn as_ref(&self) -> &elektra_sys::KDB {
+        unsafe { self.ptr.as_ref() }
+    }
+}
+
+// const ELEKTRA_ERROR_RESOURCE: &str = "C01100";
+const ELEKTRA_ERROR_OUT_OF_MEMORY: &str = "C01110";
+// const ELEKTRA_ERROR_INSTALLATION: &str = "C01200";
+const ELEKTRA_ERROR_INTERNAL: &str = "C01310";
+const ELEKTRA_ERROR_INTERFACE: &str = "C01320";
+const ELEKTRA_ERROR_PLUGIN_MISBEHAVIOR: &str = "C01330";
+const ELEKTRA_ERROR_CONFLICTING_STATE: &str = "C02000";
+const ELEKTRA_ERROR_VALIDATION_SYNTACTIC: &str = "C03100";
+const ELEKTRA_ERROR_VALIDATION_SEMANTIC: &str = "C03200";
+
+#[derive(Debug)]
+pub enum KDBError<'a> {
+    Permanent(PermanentError<'a>),
+    ConflictingState(KDBErrorWrapper<'a>),
+    Validation(ValidationError<'a>),
+}
+
+#[derive(Debug)]
+pub enum PermanentError<'a> {
+    Resource(ResourceError<'a>),
+    Logical(LogicalError<'a>),
+    Installation(KDBErrorWrapper<'a>),
+}
+
+#[derive(Debug)]
+pub enum LogicalError<'a> {
+    Internal(KDBErrorWrapper<'a>),
+    Interface(KDBErrorWrapper<'a>),
+    PluginMisbehavior(KDBErrorWrapper<'a>),
+}
+
+#[derive(Debug)]
+pub enum ResourceError<'a> {
+    OutOfMemory(KDBErrorWrapper<'a>),
+}
+
+#[derive(Debug)]
+pub enum ValidationError<'a> {
+    Syntactic(KDBErrorWrapper<'a>),
+    Semantic(KDBErrorWrapper<'a>),
+}
+
+#[derive(Debug)]
+pub struct KDBErrorWrapper<'a> {
+    error_key: StringKey<'a>,
+}
+
+impl<'a> KDBErrorWrapper<'a> {
+    /// Constructs a new KDBErrorWrapper from a StringKey.
+    /// Only pass keys where the metakeys error/* are set.
+    pub fn new(error_key: StringKey) -> KDBErrorWrapper {
+        KDBErrorWrapper { error_key }
+    }
+
+    // TODO: For which of these error/* can we be sure that they exist?
```

Comment:
I haven't found a description/specification for which `error/*` keys are always set. Since I'm unwrapping (panics if it doesn't exist) on the keys in the methods below, this would be good to know.

Reply:
@Piankero  can you help out here and create such a documentation?

Reply:
@PhilippGackstatter I am not 100%.
The code which sets the error metadata looks like this (example)
```
		keySetMeta(errorKey, "error", "number description  module file line function reason");
		keySetMeta(errorKey, "error/number", "C02000");
		keySetMeta(errorKey, "error/description", "Conflict");
		if (strcmp(module, "ELEKTRA_MODULE_NAME") == 0) module = "kdb";
		keySetMeta(errorKey, "error/module", module);
		keySetMeta(errorKey, "error/file", file);
		keySetMeta(errorKey, "error/line", line);
		keySetMeta(errorKey, "error/mountpoint", keyName(errorKey));
		keySetMeta(errorKey, "error/configfile", keyString(errorKey));
```
file and line get injected via macros and are always present.
It might be possible that configfile is empty for some errors but it should also be set in theory.

Reply:
Thank you! I'll leave it as is then, since all of the keys I'm retrieving should be set.

Reply:
@Piankero it would be good if we get a definitive answer here. Please also make sure it is documented somewhere.

Reply:
I have looked through the majority of the commands in `src/tools/kdb` and various other places and they mostly set the file `errorKey.setString (file)`. But `kdb check` for example does not set it. Also it is easily possible that some refactorings or new command forget to set the file in the errorKey so you should not count on it.

Reply:
@markus2330 is there a branch or already existing tutorial where this information fits?

Reply:
Then I'll return an empty string if the key is missing. Still better than exiting the program.

Reply:
I am only talking about the `error/configfile`. It can happen that the value there is simply empty.

Reply:
@Piankero wrote:
> @markus2330 is there a branch or already existing tutorial where this information fits?

Not really, the error information is currently very much scattered around. A new tutorial would be needed.

Reply:
> Also it is easily possible that some refactorings or new command forget

To forget something would be only a bug. But soon there will be non-file-based storage plugins which obviously cannot provide a useful configfile.